src/docs/development/add-to-app/ios/add-flutter-screen.md

```diff
@@ -6,77 +6,291 @@ description: Learn how to add a single Flutter screen to your existing iOS app.
+  and the `FlutterViewController` attaches to a `FlutterEngine` to pass UIKit
+  input events into Flutter and to display frames rendered by the `FlutterEngine`.
+{{site.alert.end}}
 
-In `AppDelegate.h`:
+The `FlutterEngine` may have the same lifespan as your `FlutterViewController`
+or outlive your `FlutterViewController`.
 
-```objective-c
+{{site.alert.tip}}
+It's generally recommended pre-warm a long-lived `FlutterEngine` for your
+application. This way,
+
+- Flutter screens will render their first frames faster.
+- Your Flutter and Dart state could outlive one `FlutterViewController`.
+- You and your plugins can interact with Flutter and your Dart logic before
+  showing UI.
+{{site.alert.end}}
+
+See [Loading sequence and performance](/docs/development/add-to-app/performance)
+for more analysis on the latency and memory trade-offs of pre-warming an engine.
+
+### Create a FlutterEngine
+
+The proper place to do this is specific to your host app. As an example, we will
+demonstrate creating a `FlutterEngine`, exposed as a property, on app startup in
+the app delegate.
+
+<ul class="nav nav-tabs" id="engine-language" role="tablist">
+  <li class="nav-item">
+    <a class="nav-link active" id="engine-objc" href="#engine-objc-tab" role="tab" aria-controls="engine-objc" aria-selected="true">Objective-C</a>
+  </li>
+  <li class="nav-item">
+    <a class="nav-link" id="engine-swift" href="#engine-swift-tab" role="tab" aria-controls="engine-swift" aria-selected="false">Swift</a>
+  </li>
+</ul>
+
+<div class="tab-content"> {% comment %} FlutterEngine language tab start {% endcomment -%}
+
+<div class="tab-pane active" id="engine-objc-tab" role="tabpanel" aria-labelledby="engine-objc-tab" markdown="1">
+
+**In `AppDelegate.h`:**
+
+<?code-excerpt "AppDelegate.h" title?>
+```objectivec
 @import UIKit;
 @import Flutter;
 
-@interface AppDelegate : FlutterAppDelegate
+@interface AppDelegate : FlutterAppDelegate // More on the FlutterAppDelegate below.
 @property (nonatomic,strong) FlutterEngine *flutterEngine;
 @end
 ```
 
-This allows `AppDelegate.m` to be simple, unless your host app needs to override
-other methods here:
+**In `AppDelegate.m`:**
 
-```objective-c
-#import <FlutterPluginRegistrant/GeneratedPluginRegistrant.h> // Only if you have Flutter Plugins
+<?code-excerpt "AppDelegate.m" title?>
+```objectivec
+#import <FlutterPluginRegistrant/GeneratedPluginRegistrant.h> // Used to connect plugins.
 
 #import "AppDelegate.h"
 
 @implementation AppDelegate
 
-// This override can be omitted if you do not have any Flutter Plugins.
 - (BOOL)application:(UIApplication *)application
     didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
-  self.flutterEngine = [[FlutterEngine alloc] initWithName:@"io.flutter" project:nil];
-  [self.flutterEngine runWithEntrypoint:nil];
+  self.flutterEngine = [[FlutterEngine alloc] initWithName:@"my flutter engine"];
+  // Runs the default Dart entrypoint with a default Flutter route.
```

Comment:
I think we should avoid mentioning the `default Flutter route` here. If anything, we should outline how to use platform channels to communicate with the Dart side to set up any non-default state.

Reply:
We do have a default flutter route on the flutter side https://api.flutter.dev/flutter/widgets/Navigator/defaultRouteName-constant.html

Reply:
It exists but it causes much confusion and frequently doesn't work as expected

Reply:
Which part specifically are you commenting on? 

The fact that there's a cascade going on in https://api.flutter.dev/flutter/material/MaterialApp-class.html and the top level navigator behavior? It's indeed confusing but it's just how Flutter works at the moment. I don't think it should ever not work as expected. This is a mostly orthogonal concept to the timing of calls to setInitialRoute.

Reply:
It's confusing becuase you can only set it at one point in the lifecycle of starting the engine, and you lose access to that completely if you set it up in certain ways, and if you call it too late it just silently fails, and there's no good way to fix any of those problems with the API so if you open a bug about it you'll be told you should be doing it differently anyway.

Reply:
Ya, the mechanism for changing the default route is confusing as mentioned below. But it doesn't change the fact that there is a default route when you don't change it.